<p>
    Self-assignments, where a variable is assigned to itself (x = x), are redundant and do not alter the state of the variable.
    They can indicate incomplete or erroneous logic and make the code harder to read and maintain.
</p>

<h2>Noncompliant Code Example</h2>

<pre>
public function main() {
   int x = 5;
   x = x;
}
</pre>
