<p>
    Conditions which always evaluate to the same value don't do any meaningful computation.
    They increase code complexity, reduce the code readability, and potentially hide logical errors.
</p>

<h2>Noncompliant Code Example</h2>

<pre>
public function main() {
   int a = x % 1; // always evaluates to zero
}
</pre>
