<p>
    When 'checkpanic' is used, the program terminates abruptly with a panic unless itâ€™s handled explicitly along the call stack.
</p>

<h2>Noncompliant Code Example</h2>

<pre>
public function checkResult() {
    json result = checkpanic getResult();

    // ...
}

public function getResult() returns json|error {
    // ...
}
</pre>

<h2>Compliant Solution</h2>

<p>
    Check and handle the error explicitly.
</p>
<pre>
public function checkResult() {
    json|error result = getResult();

    if result is error {
        // handle error
    }

    // ...
}

public function getResult() returns json|error {
    // ...
}
</pre>

<p>
    Make use of the 'check' keyword, which immediately returns the error or transfers control to an on-fail block, in contrast to
    'checkpanic' and panicking if an expression or action evaluates to an error.
</p>
<pre>
public function checkResult() returns error? {
    json result = check getResult();
}

public function getResult() returns json|error {
    // ...
}
</pre>
