<p>
    Class methods can be isolated. An isolated method is the same as an isolated function with self treated as a parameter.
    A non-isolated method will not be called concurrently. Only isolated methods are called concurrently given that
    they are guaranteed to be safe if the arguments are also safe. To allow being called concurrently,
    a public method should be marked as isolated.
</p>

<h2>Noncompliant Code Example</h2>

<pre>
class EvenNumber {
    int i = 1;

    public function generate() returns int {
        lock {
            return self.i * 2;
        }
    }
}
</pre>

<h2>Compliant Solution</h2>

<p>
    Mark public methods as isolated to ensure the method can be called concurrently.
</p>
<pre>
class EvenNumber {
    int i = 1;

    public isolated function generate() returns int {
        lock {
            return self.i * 2;
        }
    }
}
</pre>
