<p>
    An HTTP resource is safe when used for read-only operations like GET, HEAD, or OPTIONS.
    An unsafe HTTP resource is used to alter the state of an application, such as modifying the user’s profile on a web application.
</p>
<p>
    Unsafe HTTP resources include POST, PUT, and DELETE.
</p>
<p>
    Enabling both safe and insecure HTTP resources to execute a particular operation on a web application may
    compromise its security; for instance, CSRF protections are typically designed to safeguard operations executed by
    insecure HTTP resources.
</p>

<h2>Noncompliant Code Example</h2>

<pre>
listener http:Listener endpoint = new (8080);

service / on endpoint {
    // Sensitive: by default all HTTP methods are allowed
    resource function default deleteRequest(http:Request clientRequest, string username) returns string {
        // state of the application will be changed here
    }
}
</pre>

<h2>Compliant Solution</h2>

<p>
    For every resource in an application, it’s crucial to explicitly define the type of the HTTP resource,
    ensuring that safe resources are exclusively used for read-only operations.
</p>
<pre>
service / on endpoint {
    resource function delete deleteRequest(http:Request clientRequest, string username) returns string {
        // state of the application will be changed here
    }
}
</pre>
