<p>
    A non-isolated function will not be called concurrently. Only isolated functions are called concurrently given that
    they are guaranteed to be safe if the arguments are also safe. To allow being called concurrently,
    a public function should be marked as isolated.
</p>

<h2>Noncompliant Code Example</h2>

<pre>
public function helperFunction() {
   // isolated code
}
</pre>

<h2>Compliant Solution</h2>

<p>
    Mark public functions as isolated to ensure the function can be called concurrently.
</p>
<pre>
public isolated function helperFunction() {
   // isolated code
}
</pre>
