<p>
    Conditions that are always true don't do any meaningful computation. They increase code complexity,
    reduce the code readability and potentially hide logical errors.
</p>

<h2>Noncompliant Code Example</h2>

<pre>
public function main() {
   int a = 1
   boolean b = a <= int:MAX_VALUE
}
</pre>
